op-node/p2p/gossip.go

```diff
@@ -309,26 +325,58 @@ func BuildBlocksValidator(log log.Logger, cfg *rollup.Config, runCfg GossipRunti
 			return pubsub.ValidationReject
 		}
 
 		// [REJECT] if a V1 Block has withdrawals
-		if blockVersion == eth.BlockV1 && payload.Withdrawals != nil {
+		if !blockVersion.HasWithdrawals() && payload.Withdrawals != nil {
 			log.Warn("payload is on v1 topic, but has withdrawals", "bad_hash", payload.BlockHash.String())
 			return pubsub.ValidationReject
 		}
 
 		// [REJECT] if a V2 Block does not have withdrawals
-		if blockVersion == eth.BlockV2 && payload.Withdrawals == nil {
-			log.Warn("payload is on v2 topic, but does not have withdrawals", "bad_hash", payload.BlockHash.String())
+		if blockVersion.HasWithdrawals() && payload.Withdrawals == nil {
+			log.Warn("payload is on v2/v3 topic, but does not have withdrawals", "bad_hash", payload.BlockHash.String())
 			return pubsub.ValidationReject
 		}
 
 		// [REJECT] if a V2 Block has non-empty withdrawals
-		if blockVersion == eth.BlockV2 && len(*payload.Withdrawals) != 0 {
-			log.Warn("payload is on v2 topic, but has non-empty withdrawals", "bad_hash", payload.BlockHash.String(), "withdrawal_count", len(*payload.Withdrawals))
+		if blockVersion.HasWithdrawals() && len(*payload.Withdrawals) != 0 {
+			log.Warn("payload is on v2/v3 topic, but has non-empty withdrawals", "bad_hash", payload.BlockHash.String(), "withdrawal_count", len(*payload.Withdrawals))
+			return pubsub.ValidationReject
+		}
+
+		// [REJECT] if the block is on a topic <= V2 and has a blob gas value set
+		if !blockVersion.HasBlobProperties() && payload.BlobGasUsed != nil {
+			log.Warn("payload is on v1/v2 topic, but has blob gas used", "bad_hash", payload.BlockHash.String())
+			return pubsub.ValidationReject
+		}
+
+		// [REJECT] if the block is on a topic <= V2 and has an excess blob gas value set
+		if !blockVersion.HasBlobProperties() && payload.ExcessBlobGas != nil {
+			log.Warn("payload is on v1/v2 topic, but has excess blob gas", "bad_hash", payload.BlockHash.String())
+			return pubsub.ValidationReject
+		}
+
+		if blockVersion.HasBlobProperties() {
+			// [REJECT] if the block is on a topic >= V3 and has a blob gas used value that is not zero
+			if payload.BlobGasUsed == nil || (payload.BlobGasUsed != nil && *payload.BlobGasUsed != 0) {
+				log.Warn("payload is on v3 topic, but has non-zero blob gas used", "bad_hash", payload.BlockHash.String(), "blob_gas_used", payload.BlobGasUsed)
+				return pubsub.ValidationReject
+			}
+
+			// [REJECT] if the block is on a topic >= V3 and has an excess blob gas value that is not zero
+			if payload.ExcessBlobGas == nil || (payload.ExcessBlobGas != nil && *payload.ExcessBlobGas != 0) {
+				log.Warn("payload is on v3 topic, but has non-zero excess blob gas", "bad_hash", payload.BlockHash.String(), "excess_blob_gas", payload.ExcessBlobGas)
+				return pubsub.ValidationReject
+			}
+		}
+
+		// [REJECT] if the block is on a topic >= V3 and the parent beacon block root is nil
+		if blockVersion == eth.BlockV3 && envelope.ParentBeaconBlockRoot == nil {
```

Comment:
Slightly suspect that the comment says `>= V3` and the code says `== V3`. Maybe we should have a `HasParentBeaconBlockRoot` method on the block version like we do for the other blobs and withdrawals?